academy/ibc-dev/ibcapp_steps.md

```diff
@@ -0,0 +1,425 @@
+
+```bash
+modify app/app.go
+modify proto/leaderboard/genesis.proto
+create proto/leaderboard/packet.proto
+modify testutil/keeper/leaderboard.go
+modify x/leaderboard/genesis.go
+create x/leaderboard/module_ibc.go
+create x/leaderboard/types/events_ibc.go
+modify x/leaderboard/types/genesis.go
+modify x/leaderboard/types/keys.go
+```
+
+To have a more detailed view, we can now compare both versions with a git diff.
+
+## IBC application module requirements
+
+What does Ignite CLI do behind the scenes when creating an IBC module for us? What do we need to implement if we want to upgrade a regular custom application module to an IBC enabled module?
+
+The required steps to implement can be found in the [IBC go docs](https://ibc.cosmos.network/main/ibc/apps/apps.html). We find there:
+
+<Highlightbox type="info">
+
+**To have your module interact over IBC you must:**
+
+- implement the `IBCModule` interface, i.e.:
+  - channel (opening) handshake callbacks
+  - channel closing handshake callbacks
+  - packet callbacks
+- bind to a port(s)
+- add keeper methods
+- define your own packet data and acknowledgement structs as well as how to encode/decode them
+- add a route to the IBC router
+
+</Highlightbox>
+
+Now let's take a look at the *git diff* and see if we are able to recognize the steps listed above.
+
+### Implementing the `IBCModule` interface
+
+**NOTE:** For a full explanation, please visit the [IBC go docs](https://ibc.cosmos.network/main/ibc/apps/ibcmodule.html).
+
+The Cosmos SDK expects all IBC modules to implement the [`IBCModule`
+interface](https://github.com/cosmos/ibc-go/tree/main/modules/core/05-port/types/module.go). This interface contains all of the callbacks IBC expects modules to implement. This includes callbacks related to: 
+- channel handshake (`OnChanOpenInit`, `OnChanOpenTry`, `OncChanOpenAck`, `OnChanOpenConfirm`)
+- channel closing (`OnChanCloseInit`, `OnChanCloseConfirm`)
+- packets (`OnRecvPacket`, `OnAcknowledgementPacket` and `OnTimeoutPacket`).
+
+Ignite CLI implements this in the file `x/leaderboard/module_ibc.go`.
+
+//TODO: add collapsible window??
+
+Additionally, in the `module.go` file, add the following line (and the corresponding import):
+
+```diff
+var (
+    _ module.AppModule      = AppModule{}
+    _ module.AppModuleBasic = AppModuleBasic{}
+    // Add this line
++    _ porttypes.IBCModule   = IBCModule{}
+)
+```
+
+#### Channel handshake version negotiation
+
+Application modules are expected to verify versioning used during the channel handshake procedure.
+
+- `OnChanOpenInit` will verify that the relayer-chosen parameters
+  are valid and perform any custom `INIT` logic.
+  It may return an error if the chosen parameters are invalid
+  in which case the handshake is aborted.
+  If the provided version string is non-empty, `OnChanOpenInit` should return
+  the version string if valid or an error if the provided version is invalid.
```

Comment:
maybe indicate that the channel version string can also include metadata that we use to indicate attributes of the channel we are supporting, like applicable middleware and the underlying app version. 

examples: 

https://github.com/cosmos/ibc-go/blob/995b647381b909e9d6065d6c21004f18fab37f55/modules/apps/29-fee/types/metadata.pb.go#L28

https://github.com/cosmos/ibc-go/blob/995b647381b909e9d6065d6c21004f18fab37f55/modules/apps/transfer/types/keys.go#L12